### PR TITLE
Submission2

### PR DESCRIPTION
**Pull request by jatellezs**

### CODE REVIEW

File: block_model.py [hunk 1]
```diff
@@ -20,18 +24,62 @@ def __eq__(self, other):
...
+
+    def reBlock(self, rx, ry, rz):
+        if invalidParameters([rx, ry, rz]):
+            return "Invalid parameters"
+        x = 0
+        y = 0
+        z = 0
+        for block in self.blocks:
+            try:
+                if int(block.getValue("x")) > x:
```
Comment (vreyn) [inline block_model.py:41]: You shold rethink a more robust solution that doesnt depend on this two possible names for the index variables, and doesnt involve writing this try-except blocks

File: block_model.py [hunk 2]
```diff
@@ -20,18 +24,62 @@ def __eq__(self, other):
...
+        else:
+            return block
+
+    def reBlock(self, rx, ry, rz):
+        if invalidParameters([rx, ry, rz]):
+            return "Invalid parameters"
+        x = 0
+        y = 0
+        z = 0
+        for block in self.blocks:
```
Comment (vreyn) [inline block_model.py:39]: You are getting the max X, Y and Z values here, you should be able to do something like, for each variable:
`max(self.blocks, key = lambda b: b.getValue("x"))`

File: block_model.py [hunk 3]
```diff
@@ -20,18 +24,62 @@ def __eq__(self, other):
...
+        new_blocks = emptyblocks(new_x, new_y, new_z)
+        count_x = 0
+        count_y = 0
+        count_z = 0
+        for i in range(0, x, rx):
+            count_y = 0
+            for j in range(0, y, ry):
+                count_z = 0
+                for k in range(0, z, rz):
+                    new_blocks[count_x][count_y][count_z] = reblockArroundBlocks(self, i, j, k, rx, ry, rz)
```
Comment (vreyn) [inline block_model.py:79]: Very nice that you extracted this to a separate function 👍

File: reblock_functions.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                        modes = [v]
+                        mode_max = values.count(v)
+                    elif values.count(v) == mode_max and v not in modes:
+                        modes.append(v)
+                mode = modes[random.randint(0, len(modes) - 1)]
+                new_values.append(mode)
+                #print(mode)
+    return Block(columns, blocks[0].mass, blocks[0].minerals, new_values, blocks[0].classification)
+
+def reblockArroundBlocks(blockmodel, block_x, block_y, block_z, reblock_x, reblock_y, reblock_z):
```
Comment (vreyn) [inline reblock_functions.py:74]: typo Arround -> Around

File: test_block_model.py [hunk 5]
```diff
@@ -1,24 +1,170 @@
...
+        classification = [0, 0, 0, 0, 0, 1, 1, 1]
+        block_1 = Block(columns, mass, minerals, [1, 2, 5, 4, 10, 5, 6, 2], classification)
+        block_2 = Block(columns, mass, minerals, [2, 2, 5, 5, 10, 4, 7, 1], classification)
+        block_map = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
+        block_map['2']['5']['4'] = block_1
+        block_map['2']['5']['5'] = block_2
+        block_model = BlockModel(columns, [block_1, block_2], block_map)
+        self.assertEqual(block_model.getBlock(2, 3, 5), "Block does not exist")
+
+    def test_reblock_when_valid_parameters(self):
```
Comment (vreyn) [inline test_block_model.py:62]: You should more specific cases for testing reblock. The idea is to get specific information when a test fail, so the more tests for specific internal logic of a method, the better. For example:
- You should have a separate test that checks only if continous attributes reblocking works, another for mass proportional, and another for categorical. You can even use the same test blockModel you have here, but then assert only the specific values of the properties you want to check. That way, if one of this fails, you know the problem is with that type of variable.
- Other improtant test cases for reblock, is how the reblock factor is used. For example what happens if the reblock factor is not divisor of the block model size. You should have different tests for this for each of the dimensions (x,y,z), again, with the goal of giving you more specific information.
In general a good skill o acquire is to try to define many tests that test different specific case for the algorithm. This will help you find possible bugs, but also help you reason about the algorithm and find edge cases where it can fails.
